### ISSUE TITLE
Airplay devices are going into sleep and have to be reactivated by restarting slimproto and airplay player providers

### ISSUE THREAD

**Issue by poerlemans** [music-assistant/hass-music-assistant#1247]
### What version of Music Assistant has the issue?

2.0.0b30

### The problem

I operate a system with both airplay devices and Logitech squeezebox devices. So I have activated both the airplay and slimproto player providers.
As reported by others earlier the connection between music assistant and the squeezebox devices has to be improved. At the moment a lot of slimproto errors are found in the logs.

What is the issue?
After activating the airplay and slimproto player providers it is possible to play music without problems (via airplay on Denon receivers). When I stop the playing and shutoff the receiver it is possible to start again playing music, but when I wait too long the players are coming into a kind of sleep mode. It is possible to reactivate the players by disable both player providers and than enabling them again. 

### How to reproduce

See the description above

### Relevant log output

When players are going into sleep mode (I hope that this is the relevant information!):

2023-05-13 12:20:47 WARNING asyncio  -- Executing <Task finished name='Task-50290' coro=<SlimprotoProvider._create_client() done, defined at /usr/local/lib/python3.11/site-packages/music_assistant/server/providers/slimproto/__init__.py:149> result=None created at /usr/local/lib/python3.11/asyncio/streams.py:249> took 0.253 seconds
2023-05-13 12:21:27 WARNING asyncio  -- Executing <Task finished name='Task-55569' coro=<SlimprotoProvider._create_client() done, defined at /usr/local/lib/python3.11/site-packages/music_assistant/server/providers/slimproto/__init__.py:149> result=None created at /usr/local/lib/python3.11/asyncio/streams.py:249> took 0.259 seconds
2023-05-13 12:33:55 WARNING asyncio  -- Executing <Task finished name='Task-65054' coro=<SlimprotoProvider._create_client() done, defined at /usr/local/lib/python3.11/site-packages/music_assistant/server/providers/slimproto/__init__.py:149> result=None created at /usr/local/lib/python3.11/asyncio/streams.py:249> took 0.273 seconds
2023-05-13 12:34:30 WARNING asyncio  -- Executing <Task finished name='Task-69771' coro=<SlimClient._process_helo() done, defined at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:504> result=None created at /usr/local/lib/python3.11/asyncio/tasks.py:374> took 0.305 seconds
2023-05-13 12:35:09 WARNING asyncio  -- Executing <Task finished name='Task-74872' coro=<SlimClient._process_helo() done, defined at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:504> result=None created at /usr/local/lib/python3.11/asyncio/tasks.py:374> took 0.308 seconds

___________________________________________________________________________________________________________________________________
The problems concerning the connection of the Logitech squeezebox players is already reported earlier. But for reasons of completeness relevant logs are as follows:

2023-05-13 12:35:30 ERROR aioslimproto.discovery  -- Error occured while trying to parse a datagram from ('192.168.1.22', 3483) - data: b'd\x00\x07M\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04 \x16O"'
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/discovery.py", line 205, in datagram_received
    dgram = Datagram.decode(data)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/discovery.py", line 42, in decode
    return ClientDiscoveryDatagram(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/discovery.py", line 63, in __init__
    assert msg[0] == "d"
           ^^^^^^^^^^^^^
AssertionError

### What version of Home Assistant Core are your running

2023.5.2

### What type of installation are you running?

Home Assistant OS

### On what type of hardware are you running?

Generic x86-64 (e.g. Intel NUC)

**Comment by marcelveldt** [comment]
I'm trying to reproduce this one while fixing other slimproto related bugs.
Does it help if you "power off" and on the player in the MA interface ? So toggle it off and on again and then start playback ?
Or does that not help either ?

**Comment by poerlemans** [comment]
When an airplay device is in a so called "sleeping mode" it is not possible to "power on" or "power off" the player in de MA interface. See next picture: The AVR-X2800 is sleeping, light greyed and cannot be powered on or off. 
![Airplay, sleeping](https://github.com/music-assistant/hass-music-assistant/assets/32535738/3e089d82-e3dc-497f-ac8f-8b57396a7699)

Most of the time all three players are in sleeping mode. The players are all three Denon-made (AVR-3313, AVR-X2800 and Home150.
I don't know if it has to do something with slimproto bugs. I have here three Squeezebox players (touch, boom and duet) and all three have connection problems. For that reason I mentioned this.

**Comment by marcelveldt** [comment]
When a player is greyed out, it is disconnected and thus unavailable. 
I have several slimproto based players and airplay (which is using slimproto as base) and this disconnection issue is not something I recognize but we do have some other slimproto bugs that are somewhat related.

Do the squeezebox players also go unavailable after some time ?

Also interesting to know if this happens all at the same time perhaps ?
So do all players go unavailable at the exact same moment

**Comment by poerlemans** [comment]
Yesterday I carried out a clean install of MAS. Until now the squeezeboxes were not discovered by MAS. The coming day I shall force discovery by trying to connect the devices with MAS. But I know now already that a smooth connection will not take place.

The moment that an airplay device is going into sleep  is very inexplicable. Sometimes all three airplay devices are in sleep, but on other moments a device is awake. Some minutes ago AVR-3313 was ago into the sleep mode. I observed the following log:

**2023-05-29 20:56:23 WARNING asyncio  -- Executing <Task finished name='Task-1185074' coro=<SlimClient._process_helo() done, defined at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:504> result=None created at /usr/local/lib/python3.11/asyncio/tasks.py:374> took 0.232 seconds**

**Comment by poerlemans** [comment]
I just tried to connect the squeezebox touch. The device is now visible in the list of players. Connection didn't succeed.
I found the following log:

**2023-05-29 21:16:45 WARNING asyncio  -- Executing <TimerHandle when=899038.402624412 SlimClient._send_heartbeat() created at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:643> took 0.235 seconds,

2023-05-29 21:17:13 WARNING asyncio  -- Executing <TimerHandle when=899066.218535322 SlimClient._send_heartbeat() created at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:643> took 0.217 seconds

2023-05-29 21:17:41 WARNING asyncio  -- Executing <Task pending name='Task-1205355' coro=<BaseSelectorEventLoop._accept_connection2() running at /usr/local/lib/python3.11/asyncio/selector_events.py:222> wait_for=<Future pending cb=[Task.task_wakeup()] created at /usr/local/lib/python3.11/asyncio/base_events.py:427> created at /usr/local/lib/python3.11/asyncio/selector_events.py:198> took 0.221 seconds,

2023-05-29 21:18:00 WARNING asyncio  -- Executing <Task finished name='Task-1210075' coro=<SlimClient._process_helo() done, defined at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:504> result=None created at /usr/local/lib/python3.11/asyncio/tasks.py:374> took 0.207 seconds,

2023-05-29 21:18:09 WARNING asyncio  -- Executing <Task finished name='Task-1212123' coro=<SlimprotoProvider._create_client() done, defined at /usr/local/lib/python3.11/site-packages/music_assistant/server/providers/slimproto/__init__.py:149> result=None created at /usr/local/lib/python3.11/asyncio/streams.py:249> took 0.218 seconds,

2023-05-29 21:21:15 WARNING asyncio  -- Executing <Handle StreamReaderProtocol.connection_made(<_SelectorSoc...e, bufsize=0>>) created at /usr/local/lib/python3.11/asyncio/selector_events.py:915> took 0.206 seconds,
2023-05-29 21:19:20 INFO  music_assistant.players  -- Player registered: 00:04:20:22:7e:5b/controller: 00:04:20:22:7e:5b
2023-05-29 21:21:53 WARNING asyncio  -- Executing <Handle StreamReaderProtocol.connection_made(<_SelectorSoc...e, bufsize=0>>) created at /usr/local/lib/python3.11/asyncio/selector_events.py:915> took 0.223 seconds**

**Comment by poerlemans** [comment]
Included a log of AVR-X2800 (thanks to the integration with Home Assistant):
![image](https://github.com/music-assistant/hass-music-assistant/assets/32535738/1973da5c-b221-4976-9be4-99b8f0fb4522)

The log in the last comment (start sleeping at 21:18:00) corresponds with the Home Assistant log:

2023-05-29 21:18:00 WARNING asyncio -- Executing <Task finished name='Task-1210075' coro=<SlimClient._process_helo() done, defined at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:504> result=None created at /usr/local/lib/python3.11/asyncio/tasks.py:374> took 0.207 seconds,

**Comment by poerlemans** [comment]
Boom player was connected with mysqueezebox.com.
After disconnecting the Boom player is discovered by MAS. But after discovery same kind of logs:

2023-05-30 12:10:37 INFO  music_assistant.players  -- Player registered: 00:04:20:1e:6e:7f/boom: 00:04:20:1e:6e:7f

2023-05-30 12:12:11 ERROR aioslimproto.discovery  -- Error occured while trying to parse a datagram from ('192.168.1.25', 3483) - data: b'd\x00\n9\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04 \x1en\x7f'
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/discovery.py", line 205, in datagram_received
    dgram = Datagram.decode(data)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/discovery.py", line 42, in decode
    return ClientDiscoveryDatagram(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/discovery.py", line 63, in __init__
    assert msg[0] == "d"
^^^^^^^^^^^^^
AssertionError

**Comment by marcelveldt** [comment]
Please create a separate issue for the squeezebox players to keep my mind somewhat sane :-) Thanks!
As for the issue with the Airplay devices, I think I reproduced it and have a possible fix on the way, so please test it with beta33 that should arrive within a few days.

**Comment by marcelveldt** [comment]
Can you give beta36 a try ? I think/hope this is now fixed!

**Comment by OzGav** [comment]
Closing assumed fixed pending further info

**Comment by poerlemans** [comment]
I tested the airplay connection problems with beta 38.
Problem still exists with my Denon AVR and Denon home speaker. However phenomenon is very irregular. I was thinking it might have to do with my internal network, but I have the impression that network is oke (also with ethernet cable connections connection problems are occuring).
It's difficult to find relevant logs, but maybe next logs can help:

On 21:34:52 airplay connection of Denon AVR3313 was lost.
Relevant logs might be (time is not corresponding exactly):

2023-06-21 21:34:21 ERROR asyncio  -- Task exception was never retrieved
future: <Task finished name='Task-185385' coro=<SlimClient._socket_reader() done, defined at /usr/local/lib/python3.11/site-packages/aioslimproto/client.py:589> exception=ConnectionResetError(104, 'Connection reset by peer') created at /usr/local/lib/python3.11/asyncio/tasks.py:374>
source_traceback: Object created at (most recent call last):
  File "/usr/local/bin/mass", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.11/site-packages/music_assistant/__main__.py", line 108, in main
    run(
  File "/usr/local/lib/python3.11/site-packages/aiorun.py", line 273, in run
    loop.run_forever()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 607, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 1914, in _run_once
    handle._run()
  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/providers/slimproto/__init__.py", line 270, in _create_client
    SlimClient(reader, writer, client_callback)
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/client.py", line 241, in __init__
    self._reader_task = create_task(self._socket_reader())
  File "/usr/local/lib/python3.11/asyncio/tasks.py", line 374, in create_task
    task = loop.create_task(coro)
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aioslimproto/client.py", line 596, in _socket_reader
    data = await self._reader.read(64)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 690, in read
    await self._wait_for_data('read')
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 522, in _wait_for_data
    await self._waiter
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 995, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ConnectionResetError: [Errno 104] Connection reset by peer

**Comment by OzGav** [comment]
That error appears to be related to slimproto not airplay?

**Comment by poerlemans** [comment]
No, no, my squeezebox devices are running fine now. But connection problems in airplay are still occuring.

**Comment by poerlemans** [comment]
The airplay connection problem with my Denon AVR-devices is a very mysterious one.
Unfortunately, no logs are shown when such a connection problem starts.
I have an example included:
![image](https://github.com/music-assistant/hass-music-assistant/assets/32535738/1a0d260d-1e14-439a-bfc4-0374f13d9364)

In above figure you can observe that connection is no problem at all until 17:54:54 h. Before that moment there wasn't any connection problem. I have set the logs of the airplay player provider on debug, but that didn't result in any log on 17:54:54 h.
At about 16:50 h I upgraded the music assistant server from b38 to b39. Then I tested the airplay provider on my AVR-3313 for a short time and then shut-off the player.
Now (19:20 h) there is still no connection.

I don't understand this behaviour at all.

**Comment by poerlemans** [comment]
@OzGav 
What shall I do: start a new issue or
Can you put this issue on the list of open issues?

**Comment by OzGav** [comment]
Because of the way multicast is used to detect these players it could be a network problem…

**Comment by poerlemans** [comment]
@OzGav @marcelveldt 
I activated Multicast DNS on my Synology Router and that indeed had a positive effect on the connection issue.
I have one question concerning this: should I activate IGMP snooping or deactivate IGMP snooping?? I now have it deactivated.

But concerning airplay experiences I am not very enthousiastic:

- Some players are fluctuating very frequently from off to idle and back again.
- Just after restart the airplay player provider, a player starts smoothly and plays music, but, when I want to activate a player after some time nothing happens when touching the play-button. Also other players cannot be activated to play anymore.

**Comment by OzGav** [comment]
Not sure about IGMP snooping. If things are working ok without it then probably leave it off? 

As for the cycling of idle/off of your airplay device i have seen that as well so need to add something to the backlog so it gets looked at .
On your last point can you see any errors in the log after you try and play?

**Comment by poerlemans** [comment]
When I restart the air play provider I get a message in the log:

2023-06-29 14:17:09 INFO  music_assistant  -- Loaded player provider Airplay_

When I start successfully a player I also get a message in the logs:

2023-06-29 14:17:37 INFO  music_assistant.providers.slimproto.cli  -- Client connected on Telnet CLI

But when I start a player unsuccessfully I get no message in the logs at all. Just at this moment I have two players (Denon AVR  3313 and Denon Home 150) that are playing fine (and they both have shown the message of 2023-06-29 14:17:37) and one player (Denon AVR X2800) that cannot be put into the playing mode. But I know almost for 100% sure that after some time also the now functioning players will refuse playing.

**Comment by OzGav** [comment]
Perhaps you have a network bandwidth problem. Are he Denons connected via ethernet or wifi?

**Comment by poerlemans** [comment]
The Denon AVR's have a ethernet connection. The Denon Home is connected by Wifi.
I observe the phenomena in both ethernet connected devices and Wifi connected devices.
Did you get some idea from the logs I sent you (or have these logs nothing to do with the airplay phenomena)?

**Comment by OzGav** [comment]
I’m not the dev but they look like slimproto warnings except the last one which looks like a connection error hence why I asked about the connection type.  What is the MA server running on? You could try changing some of the advanced settings such as encryption and compression and see if that makes any difference

**Comment by marcelveldt** [comment]
Good news: I was able to reproduce this on my production setup, happened to all my airplay players, both the homepod minis and the 3rd party Canton speakers.

Bad news: Not being able to reproduce it on my development machine so far but i will keep trying.

BTW: This is another issue then the original one which imo is fixed. This is like a complete failure of the airplay integration which goes into some sort of bootloop.

**Comment by poerlemans** [comment]
This morning I upgraded to b58. I was wondering if the airplay connection problems are not present anymore.
But I observed the same phenomena as before. There are differences. In the past nothing could be observed in the logs (debug mode) but now that has changed. Below the logs at about 13:16:42 (the moment that airplay connection of "Slaapkamer" was lost:

2023-07-12 13:16:39.980 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:39.980237] PlayerThread:465 [0xacd2e0]: tick 1238452652
2023-07-12 13:16:40.093 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:40.092688] PlayerThread:465 [0xace0d0]: tick 1238452764
2023-07-12 13:16:40.138 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:40.137725] PlayerThread:465 [0xaceec0]: tick 1238452809
2023-07-12 13:16:40.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:40.980625] PlayerThread:465 [0xacd2e0]: tick 1238453652
2023-07-12 13:16:41.093 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:41.093042] PlayerThread:465 [0xace0d0]: tick 1238453765
2023-07-12 13:16:41.142 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:41.142038] PlayerThread:465 [0xaceec0]: tick 1238453814
2023-07-12 13:16:41.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:41.980973] PlayerThread:465 [0xacd2e0]: tick 1238454652
2023-07-12 13:16:42.094 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.093341] PlayerThread:465 [0xace0d0]: tick 1238454765
2023-07-12 13:16:42.142 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.142398] PlayerThread:465 [0xaceec0]: tick 1238454814

2023-07-12 13:16:42.344 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.341135] 
mDNSsearchCallback:662 [0xaceec0]: removing renderer (Slaapkamer)
2023-07-12 13:16:42.344 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.341196] slimproto_close:902 [0xa7a8d0] slimproto stop for Slaapkamer
2023-07-12 13:16:42.443 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.442766] decode_close:181 close decode
2023-07-12 13:16:42.517 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.516573] stream_close:513 [0xa7a8d0] close stream

2023-07-12 13:16:42.529 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.528730] PlayerThread:465 [0xaceec0]: tick 1238455200
2023-07-12 13:16:42.530 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.528884] DelRaopDevice:976 [0xaceec0]: Raop device stopped
2023-07-12 13:16:42.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:42.981336] PlayerThread:465 [0xacd2e0]: tick 1238455653
2023-07-12 13:16:43.094 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:43.093880] PlayerThread:465 [0xace0d0]: tick 1238455765
2023-07-12 13:16:43.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:43.981610] PlayerThread:465 [0xacd2e0]: tick 1238456653
2023-07-12 13:16:44.094 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.094228] PlayerThread:465 [0xace0d0]: tick 1238456766
2023-07-12 13:16:44.640 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.640383] process:529 [0xa75100] strm
2023-07-12 13:16:44.641 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.640449] process_strm:233 [0xa75100] strm command t
2023-07-12 13:16:44.641 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.640467] sendSTAT:131 [0xa75100]: ms_played: 0
2023-07-12 13:16:44.641 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.640483] sendSTAT:155 [0xa75100]: STAT: STMt
2023-07-12 13:16:44.759 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.758741] process:529 [0xa77ce8] strm
2023-07-12 13:16:44.759 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.758828] process_strm:233 [0xa77ce8] strm command t
2023-07-12 13:16:44.760 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.758848] sendSTAT:131 [0xa77ce8]: ms_played: 0
2023-07-12 13:16:44.760 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [13:16:44.758864] sendSTAT:155 [0xa77ce8]: STAT: STMt

As already stated earlier, I have multicast DNS enabled in my router settings. 
Can put these logs more light on my problem?

**Comment by marcelveldt** [comment]
OK, that will help in narrowing it down a bit.
You can see very clearly in your log that squeeze2raop (the bridge from airplay to slimproto) disconnects from the device after no longer seeing the device on mdns.

**Comment by poerlemans** [comment]
I disabled mDNS. 
I start playing on my Denon 3313 via airplay. Just after about 5 seconds, playing stops. Just at that moment connection of my three Denon Airplay players stopped. I found the following logs:

2023-07-12 15:54:18.690 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:18.690015] PlayerThread:465 [0xaceec0]: tick 1247911362
2023-07-12 15:54:19.427 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:19.426569] PlayerThread:465 [0xacd2e0]: tick 1247912098
2023-07-12 15:54:19.516 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:19.516055] PlayerThread:465 [0xace0d0]: tick 1247912188
2023-07-12 15:54:19.690 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:19.690385] PlayerThread:465 [0xaceec0]: tick 1247912362
2023-07-12 15:54:20.427 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:20.426826] PlayerThread:465 [0xacd2e0]: tick 1247913098
2023-07-12 15:54:20.516 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:20.516341] PlayerThread:465 [0xace0d0]: tick 1247913188
2023-07-12 15:54:20.691 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:20.690657] PlayerThread:465 [0xaceec0]: tick 1247913362
2023-07-12 15:54:21.427 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:21.427176] PlayerThread:465 [0xacd2e0]: tick 1247914099
2023-07-12 15:54:21.517 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:21.516684] PlayerThread:465 [0xace0d0]: tick 1247914188
2023-07-12 15:54:21.691 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:21.691004] PlayerThread:465 [0xaceec0]: tick 1247914363
2023-07-12 15:54:22.428 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.427637] PlayerThread:465 [0xacd2e0]: tick 1247915099
2023-07-12 15:54:22.517 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.516928] PlayerThread:465 [0xace0d0]: tick 1247915188
2023-07-12 15:54:22.691 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.691194] PlayerThread:465 [0xaceec0]: tick 1247915363

2023-07-12 15:54:22.980 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977646] process:529 [0xa75100] strm
2023-07-12 15:54:22.980 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977708] process_strm:230 [0xa75100] strm command f
2023-07-12 15:54:22.980 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977725] decode_flush:196 [0xa75100]: decode flush
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977740] output_flush:331 [0xa75100]: flush output buffer
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977757] sendSTAT:131 [0xa75100]: ms_played: 0
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977773] sendSTAT:155 [0xa75100]: STAT: STMf
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977864] process:529 [0xa75100] strm
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977882] process_strm:230 [0xa75100] strm command s
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977898] process_strm:304 [0xa75100], strm s autostart: 1 transition period: 0 transition type: 0 codec: p
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977916] sendSTAT:131 [0xa75100]: ms_played: 0
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977931] sendSTAT:155 [0xa75100]: STAT: STMf
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977978] process_strm:321 [0xa75100]: set fade mode: 0
2023-07-12 15:54:22.981 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.977996] codec_open:225 codec open: 'p'
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978013] pcm_open:283 pcm size: 16 rate: 44100 chan: 2 bigendian: 0
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978028] process_strm:337 [0xa75100]: codec:p, ch:2, s:16, r:44100
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978136] connect_socket:161 [0xa75100] connecting to 192.168.1.188:8096
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978165] connect_socket:175 [0xa75100] set SO_RCVBUF at 131072 bytes
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978366] PlayerThread:465 [0xacd2e0]: tick 1247915650
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978446] PlayerThread:561 [0xacd2e0]: raop connecting ...
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978448] stream_sock:602 [0xa75100] header: GET /aa:aa:bb:4f:67:0f/single/57c035bd72fb4909a82a56bacb85a91b.pcm?seek_position=26&fade_in=1 HTTP/1.0
2023-07-12 15:54:22.982 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Host: 192.168.1.188:8096
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Connection: close
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Accept: */*
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Cache-Control: no-cache
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] User-Agent: VLC/3.0.9 LibVLC/3.0.9
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Range: bytes=0-
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] 
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] 
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978491] sendSTAT:131 [0xa75100]: ms_played: 0
2023-07-12 15:54:22.983 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.978509] sendSTAT:155 [0xa75100]: STAT: STMc
2023-07-12 15:54:22.986 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:22.985679] process:533 [0xa75100] unhandled visu
2023-07-12 15:54:23.013 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:23.012701] process:533 [0xa75100] unhandled grfe
2023-07-12 15:54:23.099 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:23.062884] stream_thread:352 [0xa75100] headers: len: 337
2023-07-12 15:54:23.101 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] HTTP/1.0 200 OK
2023-07-12 15:54:23.102 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] transferMode.dlna.org: Streaming
2023-07-12 15:54:23.102 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] contentFeatures.dlna.org: DLNA.ORG_OP=00;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=0d500000000000000000000000000000
2023-07-12 15:54:23.102 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Cache-Control: no-cache
2023-07-12 15:54:23.102 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Connection: close
2023-07-12 15:54:23.103 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] icy-name: Music Assistant
2023-07-12 15:54:23.103 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] icy-pub: 0
2023-07-12 15:54:23.103 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Content-Type: audio/pcm
2023-07-12 15:54:23.103 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Date: Wed, 12 Jul 2023 13:54:23 GMT
2023-07-12 15:54:23.104 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Server: Python/3.11 aiohttp/3.8.4
2023-07-12 15:54:23.104 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] 
2023-07-12 15:54:23.104 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] 
2023-07-12 15:54:23.104 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:23.063163] sendRESP:188 [10]: RESP
2023-07-12 15:54:23.105 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:23.080225] raopcl_connect:996 [0x7fc828009190]: local interface 192.168.1.188
2023-07-12 15:54:23.105 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [15:54:23.081694] exec_request:614 [0x7fc82800d530]: ----> : write POST /auth-setup RTSP/1.0
2023-07-12 15:54:23.105 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Content-Type: application/octet-stream
2023-07-12 15:54:23.105 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Content-Length: 33
2023-07-12 15:54:23.106 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] CSeq: 1
2023-07-12 15:54:23.106 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] User-Agent: iTunes/7.6.2 (Windows; N;)
2023-07-12 15:54:23.106 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Client-Instance: 0fe574860b3f9734
2023-07-12 15:54:23.107 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] DACP-ID: 1A2B3D4EA1B2C3D4
2023-07-12 15:54:23.107 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] Active-Remote: 258428859
2023-07-12 15:54:23.112 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop]

**Comment by marcelveldt** [comment]
could you then power off/on the device still in the MA interface and start playing again ?
this log is truncated, if you go to the /log endpoint of the frontend you should get the full log printed. 

did all devices stop at once ?

**Comment by poerlemans** [comment]
[d5369777_music_assistant_beta_2023-07-12T14-24-28.357Z.log](https://github.com/music-assistant/hass-music-assistant/files/12029002/d5369777_music_assistant_beta_2023-07-12T14-24-28.357Z.log)

Included a copy of the log file (from music assistant).
I can power on the device in the MA interface, but when I do that 2 or 3 times, connection of all Airplay devices is lost.

**Comment by marcelveldt** [comment]
ok thanks. Do you by any chance hav ethe webserver access turned on ?
If so you can access the webserver at http://yourhassipaddresshere:8095/log to get the full untruncated log

**Comment by poerlemans** [comment]
I can access the logs in the webserver by moving the switch in the settings of MA.
But the logs start at 12:55:55. Shall I sent you the whole log file?

I have Home Assistant installed on a NUC as a virtual machine within Proxmox. Functions quite well.

**Comment by marcelveldt** [comment]
Yes if you can send the log that would be great. There should be no sensitive data in it but please do check it.
Also let me know the timestamp that all players lost connection, maybe we can find out some common cause.

I have another user on discord reporting more or less the same.
Maybe we should open up a new issue report as we're now talking in the issue report for the sleepy airplay devices while this is technically speaking a different issue where all slimproto devices (airplay is base don that in MA) loose connection

**Comment by poerlemans** [comment]
[Airplay logs 12 July 2023.txt](https://github.com/music-assistant/hass-music-assistant/files/12031592/Airplay.logs.12.July.2023.txt)

Included the log file (airplay provider in debug made).
I tested airplay on three devices:
Denon AVR-3313
Denon AVR-X2800
Denon Home150

Some remarks for each player:
AVR-3313
Starting 16:25:53 the player started with fast on/off interruptions
Stopped at 16:28:12 Connection lost
Starting 16:53:44 fast on/off interruptions again

AVR-X2800
16:53:44 Connected; turned off
17:39:32 Not available, disconnected
18:37:32 Connected, turned off
20:52:41 Not available, disconnected

Home 150 (Slaapkamer)
16:53:44 Connected, turned off
17:40:20 Not available, disconnected
17:40:25 Connected, turned off
17:43:06 Not available, disconnected
17:43:11 Connected, turned off
20:52:41 Not available, disconnected
20:52:46 Connected, turned off

mDNS was turned off in this timespan.

**Comment by marcelveldt** [comment]
So although we should improve our code to better handle reconnecting devices this actually is exposing some sort of network issue you have. Is everything connected with a wire? Maybe turn igmp snooping on or off on your switches? 

Did you also try them one by one?

**Comment by poerlemans** [comment]
The two AVR players are connected by ethernet cable. The Home150 by Wifi.
I can try IGMP snooping, but I understood that not every switch supports this. But I will try.

**Comment by poerlemans** [comment]
@marcelveldt : I was wondering if you got some ideas concerning this Airplay-problem from the logs I sent you.
I followed your advice and tested the airplay behaviour of only one player (AVR-3313 with ethernet connection). Results very disappointing: a lot of on/off interruptions and almost impossible to start playing. Can it have to do something with Denon-made devices??
I decided to stop for a while with testing airplay and disabled the airplay provider and enable the DLNA provider.
Now, with DLNA provider enabled, no interruptions and playing smoothly.
But in case you have ideas for improvements I can test.

**Comment by marcelveldt** [comment]
I believe there are some people running fine with Airplay on Denon but I'm not 100% certain.
Note that Airplay sends uncompressed PCM audio by default (optionally compacted with ALAC encoding) so it is a bit resource intensive. What you could try is enlarge the buffer in the player settings to like 2 seconds or something. If it works better that way, you are 100% certain its your network causing issues, for example by disrupting multicast traffic.

**Comment by poerlemans** [comment]
Enlarge the buffer to 2 seconds doesn't improve. Music stops playing after about 30 seconds. Music cannot be started up anymore.

**Comment by marcelveldt** [comment]
It almost sounds like a flaky implementation of Airplay on the Denon. 
Airplay should be rock stable and I have it working on a lot of devices. I don't know if you own any iPhones but maybe try streaming from an iPhone to see how it works ?

Another possibility could be that it needs encryption to be on in the player settings ?

Did you already to power cycle the Denon receiver ?

**Comment by poerlemans** [comment]
Yes, with my IPad the Airplay connection is running without problems. I start the Airplay connection and then start music playing e.g. via Emby.
I also tested with encryption enabled but same (disappointing) result.
I'm now going back to DLNA because that is running fine. But, as I wrote before, I will test if you found improvements.

**Comment by marcelveldt** [comment]
I'm a bit out of ideas tbh, maybe we should ask some others users on discord if someone owns Denon and how it works with Airplay. Do you have another airplay device to test this with perhaps ? Are the other slimproto based devices you have playing as expected ?

Like I said, I have tested Airplay with a lot of devices and it works very well. I'm streaming in-sync audio to 8 airplay devices at a time now as we speak, without a glitch.

I was just now scrolling through your logs again and noticed one tiny thing that might actually cause it, in any way its worth a shot. I will have to quickly adjust some code for that so please test with next version.

**Comment by poerlemans** [comment]
@marcelveldt : I tested the airplay player provider in the new (b59) server version and observed the same problems. I just made a note on discord. I hope there will come a reaction!

**Comment by marcelveldt** [comment]
Can you send me the full log again please ?

**Comment by poerlemans** [comment]
I restarted MA with Airplay player provider in debug mode.
AS soon as I observe connection problems I will send you the log file with some comments.

**Comment by marcelveldt** [comment]
How can you get out of the non-working state ?

1) Power off the player in MA interface
2) Power cycle the device itself
3) Reload airplay integration and/or slimproto integration
4) Reload entire Music Assistant server

**Comment by poerlemans** [comment]
[complete logs 21 juli 2023 no 2.txt](https://github.com/music-assistant/hass-music-assistant/files/12123496/complete.logs.21.juli.2023.no.2.txt)

Included my log file.
Only one player was activated )Denon AVR-3313). The receiver was powered on during the test.
Player was activated at 13:08:31. Connection lost at 13:29:19.

Before starting the test I power on powered off several times (within MA and on the receiver). I also restarted MA.

**Comment by JS876** [comment]
@marcelveldt @poerlemans chiming in here to say I have the same issue.

- Denon AVR S760H (Wired)
- Denon Home 150s (Wired and Wireless)

After awhile they'll go to sleep in MA and I have to restart MA to reconnect them. I thought it was my network as well, so I tried:

- HA on a separate VLAN (normal config for me)
- HA connected to main and IOT VLAN
- HA solely in IOT VLAN

I even spun up temporary instances of HA in the IOT VLAN and saw the same behavior. 

Will see if I can capture some logs around the disconnects.

**Comment by JS876** [comment]
> multicast dns (especially on Unifi) can a lot of issues. I've have experienced the most weird issues because of it and have abandoned it completely. Sometimes you are in luck, but a lot of times you are not.
> 
> Why do you have it enabled ? Can you try for a while without ?

What did you end up switching to for mDNS?

**Comment by marcelveldt** [comment]
You shouldn't be using mdns forwarders at all. Just make sure HA/MA and your devices are in the same (IoT) LAN.

**Comment by marcelveldt** [comment]
This seems to be leaning towards an Denon specific issue

**Comment by JS876** [comment]
@marcelveldt thanks for input. Will test a few additional configurations things this week. 

Additonally, I was also able to replicate this on other Airplay devices in this case the WiiM Pro (https://www.amazon.com/gp/product/B0BJDY6D1W). 

In this case I tested with 2 HA instances, to keep the noise low, I disable all other player providers except Slimproto and Airplay. 
Logging is set at the global level to Debug. 

- Instance A - attached to main VLAN + IOT VLAN. The NIC attached to the IOT VLAN was set as default for HA and then I updated MA to publish and bind to the IOT IP and interface. Airplay devices disconnected after ~1 hour (Denon and WiiM Pro)
- Instance B - running only in the IOT VLAN. No settings changed in HA or MA. Airplay devices (Denon and Wiim Pro) stay connected for ~12 hours. No disconnects listed in the Logs.

**Comment by JS876** [comment]
Thanks, I'll add those to my test cases. I'll follow up with my findings.

**Comment by poerlemans** [comment]
@marcelveldt @JS876 
I think I found the origin (for my network situation):
On my Synology RT2600 router I have the possibility to choose for "smart" Wifi connect. On my primary network (the network where all my music devices and MA are running) I had already disabled smart connect.
BUT on my iot network I had still enabled smart connect. Since I disabled smart connect (about 2 hours ago) I haven't observed any problem.
My situation:
- Fast roaming off
- mDNS off
- IGMP snooping off

**Comment by marcelveldt** [comment]
haha yeah, often those "smart" or "enhancement" options are not so smart at all. Fingers crossed that is indeed the cause for the issue!

If its indeed fixed I'd say to turn IGMP snooping back on because that should generally help with multicast traffic not flooding all ports.

**Comment by poerlemans** [comment]
Sorry guys, I was too early with my conclusions. I see the same behavior as before. I have to find something else.

**Comment by OzGav** [comment]
@JS876 Any more info from your testing?

**Comment by JS876** [comment]
@OzGav @marcelveldt apologies for the delay, running a bit behind on testing. That said, I am currently testing the latest beta and Airplay is extremely unstable. 

I am running a Unifi Network, currently configured as follows:

1. IGMP Snooping - Disabled 
2. Multicast DNS - Disabled
3. Fast roaming - Disabled (was never enabled in my network)
4. Multicast Enhancement - Disabled
5. Home Assistant - Multi-homed: Main interface in IOT VLAN, Second Interface on MAIN VLAN

With the configuration above and after restarting my UDM PRO-SE, Denon devices and HA. Airplay devices only stay connected in Music Assistant for ~2-3 minutes into playing a song. The behavior is consistent for both wired and wireless devices. 

![image](https://github.com/music-assistant/hass-music-assistant/assets/37991727/2109dcb8-2d3e-44e6-87f9-760f06a7d7a8)


I have MA set to use the interface that is on the IOT network.

![image](https://github.com/music-assistant/hass-music-assistant/assets/37991727/f095935d-ecfe-4bdc-a7b0-9a455d172382)

[MA Log File.txt](https://github.com/music-assistant/hass-music-assistant/files/12244568/MA.Log.File.txt)

**Comment by marcelveldt** [comment]
I'm starting to think this is a Denon specific issue ? I use/test a lot of Airplay devices and this issue only existed for me when my (also Unifi) network setup had some bad configuration, explainable as Airplay is based around mDNS and thus requires a proper Multicast-capable network, something Unifi likes to block by default.

As a test, could you add the device to HA, using the "apple tv" integration ?
Then start some radio stream or whatever and check if it keeps connected to HA.

Also I noticed in your log you are not using the compression option (sending lossless ALAC instead of PCM), notice that sending uncompressed PCM requires a fair amount of bandwidth so my advice would be to leave the compression option enabled, it is file compression, not sound compression so lossless.

Another setting (at player level) worth looking at is the "enable encryption" setting. Some players require that.
For example the WiiM (pro) player which is based on a Linkplay chip is known to need this setting enabled as do most other Chinese chips.

Probably not related to the original issue but still worth a shot.

I'm thinking of adding some more logging to get the latest steps of the trace as I currently do not get a grasp of the player state when it turns to unavailable. In your log I did not spot a clear disconnect or maybe that part was just truncated from the log.

**Comment by JS876** [comment]
Thanks for the ideas, will test those as well.

Just did a fresh install of MA B63:

- Disabled UPnP/DLNA and Chromecast providers to keep logging "clean"
- Set global logging to debug
- Re-added music providers
- Stopped and started MA

What I noticed so far: 
- MA defaulted ON compression for Airplay devices: Currently streaming some music and it hasn't dropped the devices yet. Will monitor and see how they behave. I'll leave a speaker playing overnight to see if connection drops at all. 
-  I also enabled encryption for the WiiM Pro - i noticed prior to this, it would drop and reconnect resulting in "ghost" devices. Will see if it does the same thing overnight. 

As for logging, I notice the same as you where it wasn't capturing the disconnects. So additionally logging around that may help identify what causes the disconnect.

**Comment by JS876** [comment]
Regarding the WiiM device: 

![image](https://github.com/music-assistant/hass-music-assistant/assets/37991727/5b7caa4c-de03-4017-84bd-9d0a0fb0eca0)

Looking at the Logs, the device is discovered and registered:
![image](https://github.com/music-assistant/hass-music-assistant/assets/37991727/0bdc4c07-d9ad-4035-9ec2-bdfc1906f064)

It is then listed as "player updated" a couple hours later. Shortly after which it gets re-registered again as a different ID:

![image](https://github.com/music-assistant/hass-music-assistant/assets/37991727/b741e721-1d6e-44bb-9418-ebe3cee01faf)
________________________________________________________________________________________________________
Regarding adding the Denon Speaker via the AppleTV integration. It's stable and hasn't disconnected so far. 

![image](https://github.com/music-assistant/hass-music-assistant/assets/37991727/421350c2-edc0-49ae-ad78-f2dd1f2317a8)

**Comment by poerlemans** [comment]
Also here my three Denon devices (AVR-3313, AVR-X2800 and Home150) are playing without any interruption when making use of the Apple TV integration within Home Assistant. I'm going from hell into heaven when transferring from Airplay within Music Assistant to Apple TV integration within Home Assistant.

**Comment by JS876** [comment]
Denon device via Apple TV integration, still going stable this morning ~17+ hours:

![image](https://github.com/music-assistant/hass-music-assistant/assets/37991727/1b2b0be4-1b11-49e1-9687-3406433cf28a)

The devices streaming audio didn't seem to drop in MA and were still available but the audio streams flaked out:

- One device was streaming via the Radio Browser - went to went noise after a couple hours. 
- One device was streaming YT Music, which stopped.

**Comment by marcelveldt** [comment]
@JS876 I believe you have a completely different issue then @poerlemans.
In his case the complete player turns unavailable. Can you please create a different issue report for your issue(s) so I can keep my sanity ? thanks!

**Comment by marcelveldt** [comment]
Ah, OK now I understand, you had the same issue as poerlemans but that issue is resolved after tweaking the network setup ?  

So that means we're one step closer because the two issues you report are actually 2 known issues:

1) When playing a very long audio stream (such as radio) at some point the sound turns into noise. Known issue, seems to be originating to the airplay library. I can reproduce this on my own setup and have it under investigation.

2) Streaming from Youtube Music sometimes refuses to play the next track . This is also a known issue that Marvin is investigating as author of the Youtube music provider. Can you confirm that it doe snot stop halfway through a track but just refuses to play the next one ? Probably there will be some YT Music related error in the log at that point.

**Comment by poerlemans** [comment]
@marcelveldt @JS876 I think the Airplay connection problems within Home Assistant of my Denon devices are not related to a network issue. I tested a whole bunch of settings. I even renewed my switches (they are now all managed) but without positive results.
Just from the moment I activated Apple TV integration within Home Assistant I haven't observed any Airplay connection problem on my Denon devices. The Apple TV integration is running now 12 hours without any hickup.

**Comment by marcelveldt** [comment]
Thanks for the additional test. 

In that case we can pinpoint the problem a bit more and it is probably related to the airplay implementation we currently use within MA (which is the same as LMS), I will have to dig into this a bit more but first I need to reproduce this somehow and so far I can't. My Airplay devices just stay available.

Unfortunately this is not going to be an easy one to fix until I can reproduce this somehow.
I might ask around if someone can loan me a Denon speaker to test with.

**Comment by marcelveldt** [comment]
BTW: For next release I'm going to add a bunch of extra logging, maybe that will give us. step by step trace of what's happening but from what I've seen last time, the airplay bridge just looses connection to the device somehow, rendering it unavailable but maybe that was just a random fluke.

**Comment by poerlemans** [comment]
Yesterday I installed (fresh) version b68. I suppose with the improved logging in debug mode. Here are my experiences.
First I have to report the following. I observe a trend in moment and duration of connection losses. It can be clarified in next log (taken from HA):

![Airplay, logs Denon AVR-3313 4 aug 2023](https://github.com/music-assistant/hass-music-assistant/assets/32535738/7c8cb299-bc70-40d5-baed-2be9aec5b856)

In the picture the connection logs of receiver AVR-3313 are given (picture from yesterday aug 4). You can see that every hour connection is established for about 2 mins. Then, after 2 mins., connection is lossed again. Only just after start-up of the MA-server the behavior is different.

Then today I changed the logging to debug for airplay. The logs are given in next file:

[Logs 5 aug tot 11 h airplay debug.txt](https://github.com/music-assistant/hass-music-assistant/files/12266718/Logs.5.aug.tot.11.h.airplay.debug.txt)

Airplay was activated on three Denon devices (AVR-3313, X2800 and Home150).

Timedata:
AVR-3313 (ethernet):
start: 9:01:28
connection lossed: 9:02:18
connected: 9:03:34
connection lossed: 10:52:28

X2800 (ethernet):
start: 9:01:28
connection lossed: 9:02:18
connected: 9:03:34
connection lossed: 10:31:26

Home150 (Wifi):
start: 9:01:28
connection lossed: 9:02:18
connected: 9:03:34
connection lossed: 9:56:41
connected: 10:54:41
connection lossed: 10:56:40

**Comment by marcelveldt** [comment]
Thanks for the additional testing and debug logs
The issue is still very clear and very much the same. The device somehow stops broadcasting itself on mdns and/or those messages do not reach our add-on:

```
2023-08-05 10:52:28.764 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [10:52:28.763154] mDNSsearchCallback:662 [0xace0d0]: removing renderer (DENON-AVR-3313)
2023-08-05 10:52:28.764 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [10:52:28.763216] slimproto_close:902 [0xa77ce8] slimproto stop for DENON AVR-3313
2023-08-05 10:52:28.872 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [10:52:28.871239] decode_close:181 close decode
2023-08-05 10:52:28.905 WARNING (MainThread) [music_assistant.providers.slimproto] Player DENON AVR-3313 disconnected
2023-08-05 10:52:28.964 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [10:52:28.963795] stream_close:513 [0xa77ce8] close stream
2023-08-05 10:52:29.026 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [10:52:29.025940] PlayerThread:465 [0xace0d0]: tick 3303401697
2023-08-05 10:52:29.028 DEBUG (MainThread) [music_assistant.providers.airplay.squeeze2raop] [10:52:29.027411] DelRaopDevice:976 [0xace0d0]: Raop device stopped
```

This is the mdns listener within the "squeeze2raop" airplay bridge so just maybe that has a timeout just set too small or a smal fluke in its mdns listener.

The zeroconf listener used in HA is actually one of the best available so it is good to see that it doe snot have the issue, that means there is hope ;-)

I'm going to dive a bit into this, tbc

**Comment by marcelveldt** [comment]
Ah! I actually found something. The default "remove" timeout is 120 seconds if a device does not report its presence with mdns and there is also a note in the technical docs that some implementations do not send the keepalive all the time... 

I'm going to extend this timeout to 1800 seconds (30 minutes) and lets see if that solves it.
This means that it will take max 30 minutes to recognize a player going offline but so be it. Eventually I can make a config toggle.

**Comment by poerlemans** [comment]
I installed b70 on top of b68.
The first impressions are more or less positive. In the b68-version the airplay devices stayed active for about 2 mins. Now, in the b70-version, they stay active for about 32 mins. But I have to gather somewhat more information because just after start-up of the server the behaviour is somewhat different than after some hours.
My intension was to do a clean install right now, but I will wait for that in order to follow the behaviour after some hours.

**Comment by poerlemans** [comment]
Situation improved in b70 with respect to b68:
In b68 connection was established during 2 mins and then connection was lost for about 58 mins.
In b70 connection is established during about 30-32 mins. Connection lost during about 28 mins.

This is illustrated in next graphical log from HA. The log starts with the states of Denon AVR-3313 airplay with b68 installed (2 mins connected combined with 58 mins connection lost).
Then at about 09:00 h this morning b70 was started. Even when the receiver is playing (at about 11:00, indicated in blue) the connection is lost at about 30-32 mins total connection time.

![Airplay, logs Denon AVR-3313 7 aug 2023](https://github.com/music-assistant/hass-music-assistant/assets/32535738/4591304d-0960-4071-aa38-80bb624089f9)

**Comment by poerlemans** [comment]
@marcelveldt  Did you already got noticed about my last experiences concerning Airplay on Denon receivers (with MA version b60)?
 See the last two comments in this issue.

**Comment by marcelveldt** [comment]
Next version of the Music Assistant server will expose a new (advanced) setting for the airplay provider allowing you to override the default behavior of the keepalive and even disable that feature entirely.

**Comment by OzGav** [comment]
Fixed in beta74. Please report any further issues

**Comment by iamthetrees** [comment]
For some reason the remove timeout feature isn't working with my Homepod mini, whether I set it to -1 or any amount of seconds. I've tried restarting the addon, the homepod and restarting home assistant, still times out after 2 minutes. I'm using beta 75.

**Comment by poerlemans** [comment]
With my Denon receiver (airplay possibilities), I'm applying a timeout of 3600 sec. In this situation connection isn't lost anymore. I'm using such a long timeout because half an hour (1800 sec) was too short.
